### DISCUSSION TITLE
엔티티 매니저에 대해 설명해주세요

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#239] [Spring Data]
.

**Comment by gyungchan-jo** [comment]
엔티티 매니저(EntityManager)는 Java Persistence API(JPA)에서 엔티티(Entity) 객체의 생명주기를 관리하고 데이터베이스와의 상호작용을 처리하는 인터페이스입니다. JPA는 자바 애플리케이션이 관계형 데이터베이스와 쉽게 상호작용할 수 있도록 표준화된 API를 제공하며, 엔티티 매니저는 이 API의 핵심 구성 요소 중 하나입니다.

엔티티 매니저의 주요 기능

- 엔티티의 생명주기 관리:
    - 영속성 컨텍스트 관리: 엔티티 매니저는 영속성 컨텍스트(persistence context)를 관리합니다. 영속성 컨텍스트는 엔티티 객체를 관리하는 일종의 캐시 역할을 합니다.
    - 생성: 새로운 엔티티를 데이터베이스에 저장할 수 있습니다.
    - 읽기: 데이터베이스에서 엔티티를 조회할 수 있습니다.
    - 업데이트: 기존 엔티티를 수정하고 변경 사항을 데이터베이스에 반영할 수 있습니다.
    - 삭제: 엔티티를 데이터베이스에서 삭제할 수 있습니다.

- 트랜잭션 관리:
    - 엔티티 매니저는 트랜잭션을 시작, 커밋, 롤백하는 기능을 제공합니다.

**Comment by nayonsoso** [comment]
엔티티 메니저에 대해 알기 위해선 영속성 컨텍스트에 대해 알아야 합니다.
`영속성 컨텍스트`는 1차 캐싱, 쓰기 지연, 변경 감지를 통해 영속 로직을 효율적으로 할 수 있게 해줍니다.
1차 캐싱을 통해 DB 까지 접근하지 않고 조회를 할 수 있게 해주고,
쓰기 지연을 통해 DB 접근을 최소화하고,
변경 감지를 통해 Entity와 DB를 자동으로 동기화해줍니다.

이렇듯 효율적인 영속 로직 수행을 위해서 엔티티는 영속성 컨텍스트에 관리되어야 합니다.
이런 작업을 하는 것이 바로  `EntityManager`입니다.
**엔티티 메니저는 엔티티의 상태를 변경하고, 영속성 컨텍스트와 상호작용함으로써 영속 로직을 수행합니다.**

#### 1️⃣ 엔티티의 상태 변경

엔티티는 영속성 컨텍스트에 대해 4가지 상태를 가질 수 있습니다.
비영속, 영속, 준영속, 삭제입니다.
엔티티 메니저는 `persist()`, `merge()`, `remove()`, `close()` 등의 함수를 호출함으로써 이 작업을 수행할 수 있습니다.

#### 2️⃣ 영속성 컨텍스트와 상호작용

엔티티 메니저는 영속성 컨텍스트와 상호작용하며 그 특성을 활용할 수 있습니다.
영속성 컨텍스트의 1차 캐시로부터 엔티티를 조회해올 수 있습니다. → `find()`
쓰기 지연 저장소에 있는 쿼리들을 flush 해서 DB와 동기화시킬 수 있습니다. → `flush()`
JPQL이나 Native query 를 쓰기지연 저장소에 적재할 수도 있습니다. → `createQuery()` or `createNativeQuery()`
